linalg/ode.hpp

```diff
@@ -12,20 +12,59 @@
 #ifndef MFEM_ODE
 #define MFEM_ODE
 
+#include "../general/communication.hpp"
 #include "../config/config.hpp"
 #include "operator.hpp"
-#include "../general/communication.hpp"
+#include <vector>
+
 
 namespace mfem
 {
 
+/// A class for storing states of previous timesteps
+class StateData
+{
+protected:
+   MemoryType mem_type;
+   int ss, smax;
+   std::vector<Vector> k;
+   Array<int> idx;
+
+public:
+   StateData () { ss = smax = 0;};
+   void  SetSize(int stages, int vsize, MemoryType mem_type);
+   inline void ShiftStages()
+   {
+      for (int i = 0; i < smax; i++) { idx[i] = (++idx[i])%smax; }
+   };
+
+   int  GetMaxSize() { return smax; };
+   int  GetSize() { return ss; };
+   void IncrementSize() { ss++; ss = std::min(ss,smax); };
+   void ResetSize() { ss = 0; };
+
+   const Vector &Get(int i);
+   void Get(int i, Vector &state);
+   void Set(int i, Vector &state);
+   void Add(Vector &state);
+
+   /// Reference access to the ith vector.
+   inline Vector & operator[](int i) { return k[idx[i]]; };
+
+   /// Const reference access to the ith vector.
+   inline const Vector &operator[](int i) const { return k[idx[i]]; };
+
+   void Print(std::ostream &out = mfem::out);
+};
+
 /// Abstract class for solving systems of ODEs: dx/dt = f(x,t)
 class ODESolver
 {
 protected:
    /// Pointer to the associated TimeDependentOperator.
    TimeDependentOperator *f;  // f(.,t) : R^n --> R^n
    MemoryType mem_type;
+   StateData state;
```

Comment:
To me, this imposes an implementation to be used by all derived classes (i.e., that all derived classes should used `StateData`)... is that the intention?  In other words, do all derived classes of `ODESolver` use `StateData`?

Reply:
Derived classes can ignore the `StateData` (or now `ODEStateData`) if there is not need for it. So their is nothing imposed. However I do agree this clutters things a bit for derived members that do not use `StateData`, however, that is a necessary consequence of the "has-a" paradigm as opposed to the "is-a paradigm."
Although I have to admit that the current implementation is not a pure "has-a" implementation.

Reply:
Small correction: Their will be  state object, but it will have zero size by default.
Perhaps instead of a direct object  a pointer (unique_ptr) should be used?

Reply:
Going to a pointer will create more clutter imo... 
The object would need to be created and destroyed, and more over the data access would be a bit clumsy.

`state[0]` would become `(*state)[0]`.

Reply:
> Derived classes can ignore the StateData (or now ODEStateData) if there is not need for it.

I personally find having unused implementation in a base class to be very confusing for a user who is deriving that class, specifically  knowing which implementations they actually have to implement and which they can ignore (e.g., Operator::height and Operator::width).

**That said**, I think with your refactor of the `ODESolver` state getter functions (e.g. `ODESolver::GetStateVector`) into an accessor function `ODESolver::GetState` for the `ODEStateData` object, it makes a lot of sense to specify the member `ODEStateData` object. If someone in the future wants to add a different implementation of `ODEStateData`, they can create an interface and add the additional implementation in a derived class. To support this flexibility, my final request is to make `ODESolver.state` private so that all changes to the `ODESolver.state` object go through the accessor function

Reply:
Agree. That the non-private part must be a debug remnant.

Reply:
Coming back to that... the AM, AB and Generalized alpha classes all use the state directly, so in the current form the State data would need to remain `protected`.

Reply:
I see.  I wanted to further understand the need for a public accessor to a protected member variable, so I implemented an interface to the `ODEStateData` class and moved the corresponding accessor in `ODESolver` to a new abstract class `ODESolverWithStates` that inherits from `ODESolver` (but does not specify an implementation of `ODEStateData`).  Because this wound up being larger than what would fit inside a GitHub "suggestion", I create the "suggestion" in a different branch and opened a PR into this branch so it would be easier to see the changes: https://github.com/mfem/mfem/pull/3993

I think this approach does a decent job maximizing both
- code reuse through composition/has-a (which is just a slight adjustment to what you've already done)
- encapsulation of state data implementation (i.e., a change to a different state data implementation for a given solver would only involve changes to that solver)
- different abstraction for solvers with history than solvers without history (so it is clear to the user which solvers implement history and which do not)

Would this approach facilitate the features you are trying to introduce?

Reply:
Thanks for your suggestion! My initial suggestion was with inheritance :), but I have to admit this is way better and more general then my initial `MultiStepSolver` class.  Yet another proof that the review process improves the final output!

I can only think of one drawback.  In generic code you need a `dynamic_cast<ODESolverWithStates*> `  before the 
`StateData` can be accessed. But I guess this is the price to keep the `ODESolver` untouched.

Would it be okay to add the following, member functions to `ODESolver`

```
virtual bool hasStateData() {return false};`

```
and `ODESolverWithStates`

```
virtual bool hasStateData() {return true};
```

This would make the statedata check a bit less prozaic, but would mean a small modification to the `ODESolver` base class.

Reply:
> My initial suggestion was with inheritance :), but I have to admit this is way better and more general then my initial MultiStepSolver class. Yet another proof that the review process improves the final output!

I do apologize for the somewhat roundabout way this returned to using inheritance.  With the initial implementation with the getters and setters for the state data still in `ODESolver`, it wasn't clear to me what polymorphic need inheritance would serve.  That all said, I complete agree on the review process improving the final result and have appreciated the interactions thus far!

> I can only think of one drawback. In generic code you need a dynamic_cast<ODESolverWithStates*> before the
StateData can be accessed.

I agree that the downcasting via `dynamic_cast` is a bit unsavory.  Ideally, the initial derived class object would not be generalized past `ODESovlerWithStates` to `ODESolver` in the inheritance tree. For a specific example, one could refactor the `CheckODE::order` method into one for solvers without states/history and one for solvers with states/history:
```c++
      double order_with_restart(ODESolverWithStates* ode_solver)
      {
         double dt_order,t,order = -1;
         Vector u(2);
         Vector error(levels);
         int steps = ti_steps;

         t = 0.0;
         dt_order = t_final/double(steps);
         u = u0;
         ode_solver->Init(*oper);
         init_hist(ode_solver,dt_order);
         ode_solver->Run(u, t, dt_order, t_final - 1e-12);
         u +=u0;
         error[0] = u.Norml2();

         mfem::out<<std::setw(12)<<"Error"
                  <<std::setw(12)<<"Ratio"
                  <<std::setw(12)<<"Order"<<std::endl;
         mfem::out<<std::setw(12)<<error[0]<<std::endl;

         std::vector<Vector> uh(ode_solver->GetState().MaxSize());
         for (int l = 1; l < levels; l++)
         {
            int lvl = static_cast<int>(pow(2,l));
            t = 0.0;
            dt_order *= 0.5;
            u = u0;
            ode_solver->Init(*oper);

            // Instead of single run command
            // Chop-up sequence with Get/Set in between
            // in order to test these routines
            for (int ti = 0; ti < steps; ti++)
            {
               ode_solver->Step(u, t, dt_order);
            }
            int nstate = ode_solver->GetState().Size();

            for (int s = 0; s < nstate; s++)
            {
               ode_solver->GetState().Get(s,uh[s]);
            }

            for (int ll = 1; ll < lvl; ll++)
            {
               // Use alternating options for setting the StateVector
               if (ll%2  == 0)
               {
                  // ode_solver->state.ResetSize();
                  for (int s = nstate - 1; s >= 0; s--)
                  {
                     ode_solver->GetState().Add(uh[s]);
                  }
               }
               else
               {
                  for (int s = 0; s < nstate; s++)
                  {
                     ode_solver->GetState().Get(s,uh[s]);
                  }
               }

               for (int ti = 0; ti < steps; ti++)
               {
                  ode_solver->Step(u, t, dt_order);
               }
               nstate = ode_solver->GetState().Size();

               for (int s = 0; s< nstate; s++)
               {
                  uh[s] = ode_solver->GetState().Get(s);
               }
            }

            u += u0;
            error[l] = u.Norml2();
            order = log(error[l-1]/error[l])/log(2);
            mfem::out<<std::setw(12)<<error[l]
                     <<std::setw(12)<<error[l-1]/error[l]
                     <<std::setw(12)<<order
                     <<std::endl;
            if (error[l] < 1e-10) { break; }
         }
         delete ode_solver;

         return order;
      }

      double order(ODESolver* ode_solver)
      {
         double dt_order,t,order = -1;
         Vector u(2);
         Vector error(levels);
         int steps = ti_steps;

         t = 0.0;
         dt_order = t_final/double(steps);
         u = u0;
         ode_solver->Init(*oper);
         ode_solver->Run(u, t, dt_order, t_final - 1e-12);
         u +=u0;
         error[0] = u.Norml2();

         mfem::out<<std::setw(12)<<"Error"
                  <<std::setw(12)<<"Ratio"
                  <<std::setw(12)<<"Order"<<std::endl;
         mfem::out<<std::setw(12)<<error[0]<<std::endl;

         for (int l = 1; l < levels; l++)
         {
            int lvl = static_cast<int>(pow(2,l));
            t = 0.0;
            dt_order *= 0.5;
            u = u0;
            ode_solver->Init(*oper);
            ode_solver->Run(u, t, dt_order, t_final-1e-12);

            u += u0;
            error[l] = u.Norml2();
            order = log(error[l-1]/error[l])/log(2);
            mfem::out<<std::setw(12)<<error[l]
                     <<std::setw(12)<<error[l-1]/error[l]
                     <<std::setw(12)<<order
                     <<std::endl;
            if (error[l] < 1e-10) { break; }
         }
         delete ode_solver;

         return order;
      }
```

While there is a little bit of copied code between the two (which could probably be cleaned up with a utility function or two), I personally find the readability greatly improved.

Reply:
> This would make the statedata check a bit less prozaic, but would mean a small modification to the ODESolver base class.

If I'm not mistaken, the statedata check would still need to be followed by a cast of the `ODESolver` pointer to a `ODESolverWithStates` pointer. Unless I'm missing something, I prefer the approach I highlighted above where the original object is not passed as an `ODESolver` if it will later be treated as an `ODESolverWithStates`.